sdk/eventhub/azure-eventhub/azure/eventhub/_eventprocessor/partition_context.py

```diff
@@ -59,24 +59,26 @@ def last_enqueued_event_properties(self):
             return get_last_enqueued_event_properties(self._last_received_event)
         return None
 
-    def update_checkpoint(self, event):
-        # type: (EventData) -> None
+    def update_checkpoint(self, event=None):
+        # type: (Optional[EventData]) -> None
         """Updates the receive checkpoint to the given events offset.
 
         :param ~azure.eventhub.EventData event: The EventData instance which contains the offset and
          sequence number information used for checkpoint.
         :rtype: None
         """
         if self._checkpoint_store:
-            checkpoint = {
-                "fully_qualified_namespace": self.fully_qualified_namespace,
-                "eventhub_name": self.eventhub_name,
-                "consumer_group": self.consumer_group,
-                "partition_id": self.partition_id,
-                "offset": event.offset,
-                "sequence_number": event.sequence_number,
-            }
-            self._checkpoint_store.update_checkpoint(checkpoint)
+            checkpoint_event = event or self._last_received_event
+            if checkpoint_event:
+                checkpoint = {
+                    "fully_qualified_namespace": self.fully_qualified_namespace,
+                    "eventhub_name": self.eventhub_name,
+                    "consumer_group": self.consumer_group,
+                    "partition_id": self.partition_id,
+                    "offset": checkpoint_event.offset,
+                    "sequence_number": checkpoint_event.sequence_number,
+                }
+                self._checkpoint_store.update_checkpoint(checkpoint)
```

Comment:
We should allow for repeat checkpointing against the same event multiple times if no new events are coming through.
This section of code wouldn't need to change, however we want `self._last_received_event` to be preserved even when we are receiving empty batches (or None).

Reply:
As per discussion, the code change is to remove "partition_context._last_received_event = None  # pylint:disable=protected-access" from the following code of EventProcessor.
```python
async def _on_event_received(
    self, partition_context: PartitionContext, event: Union[Optional[EventData], List[EventData]]
) -> None:
    if event:
        try:
            partition_context._last_received_event = event[-1]  # pylint:disable=protected-access
        except TypeError:
            partition_context._last_received_event = event  # pylint:disable=protected-access
        with self._context(event):
            await self._event_handler(partition_context, event)
    else:
        partition_context._last_received_event = None  # pylint:disable=protected-access
        await self._event_handler(partition_context, event)
```